File: #0368.largest-divisible-subset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {

        vector<int> result;
        if (nums.empty()) {
            return result;
        }

        std::sort(nums.begin(), nums.end());
        int size = nums.size();
        vector<int> dp(size, 1);
        vector<int> index(size, -1);
        int max_index = 0;
        int max_dp = 1;
        for (int i = 0; i < size; i++) {
            for (int j = i - 1; j >=0 ; j--) {
                if (nums[i] % nums[j] == 0 && dp[j] + 1 > dp[i]) {
                    dp[i] = dp[j] + 1;
                    index[i] = j;
                }
            }
            if (max_dp < dp[i]) {
                max_dp = dp[i];
                max_index = i;
            }
        }
        for (int i = max_index; i != -1; i = index[i]) {
            result.push_back(nums[i]);
        }
        return result;
    }
};

File: #0368.largest-divisible-subset.ts
function largestDivisibleSubset(nums: number[]): number[] {
    nums.sort((a, b) => a - b);
    let max = [];
    const cache = []
    for (let i=nums.length - 1; i>=0; i -= 1) {
        let bestMatch = [];
        for (let j=i + 1; j < nums.length; j += 1) {
            if ((nums[i] % nums[j] === 0 || nums[j] % nums[i] === 0) 
                && cache[j].length >= bestMatch.length) {
                bestMatch = cache[j];
            }
        }
        cache[i] = bestMatch.concat(nums[i]);
        if (cache[i].length > max.length) {
            max = cache[i];
        }
    }
    return max;
};
